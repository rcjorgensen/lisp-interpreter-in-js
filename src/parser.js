import { TokenType, isPrimitive } from "./token-type.js";
import { Token } from "./token.js";

export class Parser {
  #scanner;
  /**
   * @param {import("./scanner.js").Scanner} scanner
   */
  constructor(scanner) {
    this.#scanner = scanner;
  }

  /**
   * Parse the following grammar rule:
   * `program = { expression } .`
   */
  parseProgram() {
    const expressions = [];
    while (this.#scanner.symbol !== TokenType.EOF) {
      const expression = this.parseExpression();
      expressions.push(expression);
    }

    return expressions;
  }

  /**
   * Parse the following grammar rule:
   * `expression = ( expressions ) | identifier | literal | quotedExpression .`
   */
  parseExpression() {
    if (this.#scanner.symbol === TokenType.leftParen) {
      return this.#parseExpressions();
    } else if (isPrimitive(this.#scanner.symbol)) {
      return this.#parsePrimitive();
    } else if (this.#scanner.symbol === TokenType.apostrophe) {
      return this.#parseQuotedExpression();
    }

    throw Error(`Invalid expression ${this.#scanner.symbol.label}`);
  }

  /**
   * Parse the following grammar rule:
   * `expressions = { expression } .`
   */
  #parseExpressions() {
    this.#match(TokenType.leftParen);
    const expressions = [];

    while (this.#scanner.symbol !== TokenType.rightParen) {
      const expression = this.parseExpression();
      expressions.push(expression);
    }

    this.#match(TokenType.rightParen);

    return expressions;
  }

  #parsePrimitive() {
    const literal = this.#scanner.token;
    this.#matchCurrentSymbol();
    return literal;
  }

  /**
   * Parse the following grammar rule:
   * `quotedExpression = "'" expression .`
   */
  #parseQuotedExpression() {
    const position = this.#scanner.token.position;
    this.#match(TokenType.apostrophe);
    const expression = this.parseExpression();
    return [new Token(TokenType.quoteRW, position), expression];
  }

  /**
   * Check that the current scanner symbol is the expected symbol.
   * If it is, then advance the scanner. Otherwise, throw an exception.
   */
  #match(expectedSymbol) {
    if (this.#scanner.symbol === expectedSymbol) {
      this.#scanner.advance();
    } else {
      throw Error(
        `Expecting "${expectedSymbol}" but found "${this.#scanner.token}" instead.`,
      );
    }
  }

  /**
   * Advance the scanner. This method represents an unconditional match
   * with the current scanner symbol.
   */
  #matchCurrentSymbol() {
    this.#scanner.advance();
  }
}
