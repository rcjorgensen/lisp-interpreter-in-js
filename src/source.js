import { Position } from "./position.js";

export class Source {
  EOF = -1;

  /**
   * @type {number}
   */
  #currentChar;

  get currentChar() {
    return this.#currentChar;
  }

  #lineNumber = 1;
  #charNumber = 0;
  get charPosition() {
    return new Position(this.#lineNumber, this.#charNumber);
  }

  advance() {
    if (this.#currentChar === "\n".codePointAt(0)) {
      ++this.#lineNumber;
      this.#charNumber = 1;
    } else {
      ++this.#charNumber;
    }

    if (this.#pos >= this.#source.length) {
      this.#currentChar = this.EOF;
    } else {
      this.#currentChar = this.#source.codePointAt(this.#pos);
      this.#pos += this.#currentChar > 0xffff ? 2 : 1; // Handles surrogate pairs.
    }
  }

  #source;
  #pos = 0;

  /**
   * @param {string} source
   */
  constructor(source) {
    this.#source = source;
    this.advance();
  }
}
