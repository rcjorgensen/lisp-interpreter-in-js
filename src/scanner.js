import { Position } from "./position.js";
import { Source } from "./source.js";
import { TokenBuffer } from "./token-buffer.js";
import { Token } from "./token.js";
import { TokenType } from "./token-type.js";
import { StringBuilder } from "./string-builder.js";

export class Scanner {
  #source;
  #tokenBuffer;
  #scanBuffer = new StringBuilder(100);
  #reservedWords = [
    TokenType.quoteRW,
    TokenType.setRW,
    TokenType.defineRW,
    TokenType.ifRW,
    TokenType.lambdaRW,
    TokenType.beginRW,
    TokenType.condRW,
  ];

  /**
   * @param {Source} source
   * @param {number} k
   */
  constructor(source, k) {
    this.#source = source;
    this.#tokenBuffer = new TokenBuffer(k);

    this.#reservedWords.sort(({ label: a }, { label: b }) => {
      if (a < b) return -1;
      else if (b < a) return 1;
      return 0;
    });

    for (let i = 0; i < k; ++i) {
      this.advance();
    }
  }

  get token() {
    return this.lookahead(1);
  }

  get symbol() {
    return this.lookahead(1).symbol;
  }

  get text() {
    return this.lookahead(1).text;
  }

  get position() {
    return this.lookahead(1).position;
  }

  lookahead(i) {
    return this.#tokenBuffer.get(i - 1);
  }

  advance() {
    this.#tokenBuffer.add(this.#nextToken());
  }

  advanceTo(symbol) {
    while (this.symbol !== symbol && this.symbol !== "EOF") {
      this.advance();
    }
  }

  advanceTo(symbols) {
    while (!symbols.contains(this.symbol) && this.symbol !== "EOF") {
      this.advance();
    }
  }

  #nextToken() {
    let symbol;
    let position = new Position();
    let text = "";

    this.#skipWhitespace();

    position = this.#source.charPosition;

    if (this.#source.currentChar === this.#source.EOF) {
      symbol = TokenType.EOF;
    } else if (isLetter(this.#source.currentChar)) {
      const idString = this.#scanIdentifier();
      symbol = this.#getIdentifierSymbol(idString);

      if (symbol === TokenType.identifier) {
        text = idString;
      }
    } else if (isDigit(this.#source.currentChar)) {
      symbol = TokenType.intLiteral;
      text = this.#scanIntegerLiteral();
    } else {
      switch (this.#source.currentChar) {
        case "+": {
          symbol = TokenType.plus;
          this.#source.advance();
          break;
        }
        case "-": {
          symbol = TokenType.minus;
          this.#source.advance();
          break;
        }
        case "*": {
          symbol = TokenType.times;
          this.#source.advance();
          break;
        }
        case "/": {
          symbol = TokenType.divide;
          this.#source.advance();
          break;
        }
        case ";": {
          this.#skipComment();
          return this.#nextToken();
        }
        case "<": {
          this.#source.advance();
          if (this.#source.currentChar === "=") {
            symbol = TokenType.lessOrEqual;
            this.#source.advance();
          } else {
            symbol = TokenType.lessThan;
          }
          break;
        }
        case ">": {
          this.#source.advance();
          if (this.#source.currentChar === "=") {
            symbol = TokenType.greaterOrEqual;
            this.#source.advance();
          } else {
            symbol = TokenType.greaterThan;
          }
          break;
        }
        case "(": {
          symbol = TokenType.leftParen;
          this.#source.advance();
          break;
        }
        case ")": {
          symbol = TokenType.rightParen;
          this.#source.advance();
          break;
        }
        case "'": {
          symbol = TokenType.apostrophe;
          this.#source.advance();
          break;
        }
        case '"': {
          symbol = TokenType.stringLiteral;
          text = this.#scanStringLiteral();
          break;
        }
        default:
          throw Error(`Invalid character '${this.#source.currentChar}'`);
      }
    }

    return new Token(symbol, position, text);
  }

  #skipWhitespace() {
    while (
      this.#source.currentChar !== this.#source.EOF &&
      isWhitespace(this.#source.currentChar)
    ) {
      this.#source.advance();
    }
  }

  #skipComment() {
    while (
      this.#source.currentChar !== "\n" &&
      this.#source.currentChar !== this.#source.EOF
    ) {
      this.#source.advance();
    }

    this.#source.advance();
  }

  #scanIdentifier() {
    this.#scanBuffer.clear();

    do {
      this.#scanBuffer.append(this.#source.currentChar);
      this.#source.advance();
    } while (
      isLetterOrDigit(this.#source.currentChar) ||
      isQuestionMarkOrExclamationPoint(this.#source.currentChar)
    );

    return this.#scanBuffer.toString();
  }

  #scanIntegerLiteral() {
    this.#scanBuffer.clear();

    do {
      this.#scanBuffer.append(this.#source.currentChar);
      this.#source.advance();
    } while (isDigit(this.#source.currentChar));

    return this.#scanBuffer.toString();
  }

  #scanStringLiteral() {
    this.#scanBuffer.clear();

    let c = this.#source.currentChar;
    this.#scanBuffer.append(c);
    this.#source.advance();

    while (this.#source.currentChar !== '"') {
      c = this.#source.currentChar;
      if (c === "\\") {
        this.#scanBuffer.append(this.#scanEscapedChar());
      } else {
        this.#scanBuffer.append(c);
        this.#source.advance();
      }
    }
    c = this.#source.currentChar;
    this.#scanBuffer.append(c);
    this.#source.advance();

    return this.#scanBuffer.toString();
  }

  #scanEscapedChar() {
    // Need to save current position for error reporting.
    const backslashPosition = this.#source.charPosition;

    this.#source.advance();
    const c = this.#source.currentChar;

    this.#source.advance(); // leave source at second character following backslash

    switch (c) {
      case "t":
        return "\\t";
      case "n":
        return "\\n";
      case "r":
        return "\\r";
      case '"':
        return '\\"';
      case "'":
        return "\\'";
      case "\\":
        return "\\\\";
      default: {
        throw Error(
          `Illegal escape character '\\${c}' at ${backslashPosition}.`,
        );
      }
    }
  }

  #getIdentifierSymbol(idString) {
    return (
      searchForSymbol(
        idString,
        this.#reservedWords,
        0,
        this.#reservedWords.length,
      ) ?? TokenType.identifier
    );
  }
}

function isWhitespace(char) {
  return char.trim() === "";
}

function isLetter(char) {
  return /[a-z]/i.test(char);
}

function isDigit(char) {
  return /\d/.test(char);
}

function isLetterOrDigit(char) {
  return isLetter(char) || isDigit(char);
}

function isQuestionMarkOrExclamationPoint(char) {
  return /[?!]/.test(char);
}

/**
 * @param {string} searchString
 * @param {{ label: string}[]} symbols
 * @param {number} low
 * @param {number} high
 */
function searchForSymbol(searchString, symbols, low, high) {
  if (low >= high) {
    return null;
  }

  const mid = Math.floor((low + high) / 2);
  const midSymbol = symbols[mid];
  const label = midSymbol.label;

  if (searchString === label) {
    return midSymbol;
  } else if (searchString < label) {
    return searchForSymbol(searchString, symbols, low, mid);
  } else {
    return searchForSymbol(searchString, symbols, mid + 1, high);
  }
}
