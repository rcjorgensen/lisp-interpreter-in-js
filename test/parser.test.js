import { describe, expect, it } from "vitest";

import { Source } from "../src/source.js";
import { Scanner } from "../src/scanner.js";
import { Parser } from "../src/parser.js";

import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { readdir, readFile } from "node:fs/promises";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe("correct", async () => {
  const correctDir = join(__dirname, "..", "examples", "parser", "correct");
  const files = (await readdir(correctDir)).filter((f) => f.endsWith(".scm"));

  for (const file of files) {
    it(file, async () => {
      const filepath = join(correctDir, file);
      const source = new Source(
        await readFile(filepath, { encoding: "utf-8" }),
      );
      const scanner = new Scanner(source, 4);
      const parser = new Parser(scanner);
      const program = parser.parseProgram();

      expect(program).toMatchSnapshot();
    });
  }
});
