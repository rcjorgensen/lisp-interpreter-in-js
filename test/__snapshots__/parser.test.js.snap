// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`correct > correct-01.scm 1`] = `
[
  [
    Token {
      "position": Position {
        "charNumber": 2,
        "lineNumber": 1,
      },
      "symbol": {
        "label": "define",
      },
      "text": "",
    },
    [
      Token {
        "position": Position {
          "charNumber": 10,
          "lineNumber": 1,
        },
        "symbol": {
          "label": "Identifier",
        },
        "text": "append",
      },
      Token {
        "position": Position {
          "charNumber": 17,
          "lineNumber": 1,
        },
        "symbol": {
          "label": "Identifier",
        },
        "text": "x",
      },
      Token {
        "position": Position {
          "charNumber": 19,
          "lineNumber": 1,
        },
        "symbol": {
          "label": "Identifier",
        },
        "text": "y",
      },
    ],
    [
      Token {
        "position": Position {
          "charNumber": 4,
          "lineNumber": 2,
        },
        "symbol": {
          "label": "if",
        },
        "text": "",
      },
      [
        Token {
          "position": Position {
            "charNumber": 8,
            "lineNumber": 2,
          },
          "symbol": {
            "label": "Identifier",
          },
          "text": "null?",
        },
        Token {
          "position": Position {
            "charNumber": 14,
            "lineNumber": 2,
          },
          "symbol": {
            "label": "Identifier",
          },
          "text": "x",
        },
      ],
      Token {
        "position": Position {
          "charNumber": 7,
          "lineNumber": 3,
        },
        "symbol": {
          "label": "Identifier",
        },
        "text": "y",
      },
      [
        Token {
          "position": Position {
            "charNumber": 8,
            "lineNumber": 4,
          },
          "symbol": {
            "label": "Identifier",
          },
          "text": "cons",
        },
        [
          Token {
            "position": Position {
              "charNumber": 14,
              "lineNumber": 4,
            },
            "symbol": {
              "label": "Identifier",
            },
            "text": "car",
          },
          Token {
            "position": Position {
              "charNumber": 18,
              "lineNumber": 4,
            },
            "symbol": {
              "label": "Identifier",
            },
            "text": "x",
          },
        ],
        [
          Token {
            "position": Position {
              "charNumber": 14,
              "lineNumber": 5,
            },
            "symbol": {
              "label": "Identifier",
            },
            "text": "append",
          },
          [
            Token {
              "position": Position {
                "charNumber": 22,
                "lineNumber": 5,
              },
              "symbol": {
                "label": "Identifier",
              },
              "text": "cdr",
            },
            Token {
              "position": Position {
                "charNumber": 26,
                "lineNumber": 5,
              },
              "symbol": {
                "label": "Identifier",
              },
              "text": "x",
            },
          ],
          Token {
            "position": Position {
              "charNumber": 29,
              "lineNumber": 5,
            },
            "symbol": {
              "label": "Identifier",
            },
            "text": "y",
          },
        ],
      ],
    ],
  ],
  [
    Token {
      "position": Position {
        "charNumber": 2,
        "lineNumber": 7,
      },
      "symbol": {
        "label": "Identifier",
      },
      "text": "append",
    },
    [
      Token {
        "position": Position {
          "charNumber": 9,
          "lineNumber": 7,
        },
        "symbol": {
          "label": "quote",
        },
        "text": "",
      },
      [
        Token {
          "position": Position {
            "charNumber": 11,
            "lineNumber": 7,
          },
          "symbol": {
            "label": "Identifier",
          },
          "text": "a",
        },
        Token {
          "position": Position {
            "charNumber": 13,
            "lineNumber": 7,
          },
          "symbol": {
            "label": "Identifier",
          },
          "text": "b",
        },
        Token {
          "position": Position {
            "charNumber": 15,
            "lineNumber": 7,
          },
          "symbol": {
            "label": "Identifier",
          },
          "text": "c",
        },
      ],
    ],
    [
      Token {
        "position": Position {
          "charNumber": 18,
          "lineNumber": 7,
        },
        "symbol": {
          "label": "quote",
        },
        "text": "",
      },
      [
        Token {
          "position": Position {
            "charNumber": 20,
            "lineNumber": 7,
          },
          "symbol": {
            "label": "Identifier",
          },
          "text": "d",
        },
        Token {
          "position": Position {
            "charNumber": 22,
            "lineNumber": 7,
          },
          "symbol": {
            "label": "Identifier",
          },
          "text": "e",
        },
        Token {
          "position": Position {
            "charNumber": 24,
            "lineNumber": 7,
          },
          "symbol": {
            "label": "Identifier",
          },
          "text": "f",
        },
      ],
    ],
  ],
]
`;

exports[`correct > correct-02.scm 1`] = `
[
  Token {
    "position": Position {
      "charNumber": 1,
      "lineNumber": 1,
    },
    "symbol": {
      "label": "Integer Literal",
    },
    "text": "5",
  },
]
`;
